### PR TITLE
Travis CI bug in macOS environment #17178 workaround

### PR DESCRIPTION
**Pull request by sadrasabouri**
According to #17178 by MarcoFalke, `cd` command (using for changing directory) has some conflicts with Travis builtin command with similar name in some environments (like macOS).

In this PR :
1. I tried to solve this issue by replacing `cd` by `type cd` in `ci/test/06_script_a.sh` and `ci/test/06_script_b.sh` .
2. I added function called `safe_cd` as commented in [here](https://github.com/bitcoin/bitcoin/pull/16597/files#r336052676) to avoid duplication in code.

### CODE REVIEW

**Comment by Sjors** [comment]
Why did you close this? I think `safe_cd` is a reasonable workaround. But I would change the PR title, because it doesn't really "solve" the problem.

**Comment by sadrasabouri** [comment]
@Sjors Ok Thanks.
I'll reopen the pull request.
 what can i use instead of "solve"

**Comment by sadrasabouri** [comment]
@Sjors Ok Thanks.
I'll reopen the pull request.
what can i use instead of "solve"?

**Comment by laanwj** [comment]
I'm OK with this if it solves a problem.

But to be honest I think this is terrible. Even a basic command that has existed since the 70's like `cd` needs a workaround these days? This makes it impossible to do shell scripting sanely.

File: ci/test/06_script_a.sh [hunk 1]
```diff
@@ -6,6 +6,18 @@
 
 export LC_ALL=C.UTF-8
 
+lcd=cd
+if [[ "$OSTYPE" == "darwin"* ]]; then
+  lcd=type cd
+fi
+
+safe_cd()
```
Comment (laanwj) [inline ci/test/06_script_a.sh]: at least add a comment for this workaround here: why this is needed, and when it can be removed again

**Comment by sadrasabouri** [comment]
> I'm OK with this if it solves a problem.
> 
> But to be honest I think this is terrible. Even a basic command that has existed since the 70's like `cd` needs a workaround these days? This makes it impossible to do shell scripting sanely. Are you going to review new code for using the right `cd` variant?

Unfortunately Travis CI is in it's Betha version and has several bugs in macOS environment.
In this specific issue, i am going to use `type cd` instead of `cd` in macOS and continue using `cd` for other operating systems.
In macOS, Travis considers `cd` command as it's built-in Travis command and not as a shell scripting command :cry:.

File: ci/test/06_script_a.sh (hunk 1 above)
Comment (sadrasabouri) [reply ci/test/06_script_a.sh]: Ok, thanks.
I'll add.

**Comment by MarcoFalke** [comment]
>  In macOS, Travis considers cd command as it's built-in Travis command and not as a shell scripting command cry.

why? is there any documention available? What would happen if we forced it to use the shell built in?

**Comment by sadrasabouri** [comment]
> > In macOS, Travis considers cd command as it's built-in Travis command and not as a shell scripting command cry.
> 
> why? is there any documention available? What would happen if we forced it to use the shell built in?

I really don't know :disappointed: 
I searched for a document for this specific issue, but i didn't find one.
Actually there is some issues in Travis repository like [this](https://github.com/travis-ci/travis-ci/issues/10165) that concerns about similar problem.

How can i force Travis to use shell built in? 
Any help will be appreciated :hand:

**Comment by MarcoFalke** [comment]
So what is the output of `which cd`, `type cd` and `command -v cd`?

**Comment by ryanofsky** [comment]
> How can i force Travis to use shell built in?

We should use `which` / `type` as marco suggested to actually debug this and figure out the problem so it can be fixed upstream. But replacing `cd` with `command cd` would force using the built in.

**Comment by sadrasabouri** [comment]
> > How can i force Travis to use shell built in?
> 
> We should use `which` / `type` as marco suggested to actually debug this and figure out the problem so it can be fixed upstream. But replacing `cd` with `command cd` would force using the built in.

Thank you for your help :+1: 
So i am going to use `command cd` instead of `cd` in all OSs(including macOs and ...) and use `type cd` before any usage of `command cd` in code for future debug in upstream.
Did i understand you right?

**Comment by sadrasabouri** [comment]
> > How can i force Travis to use shell built in?
> 
> We should use `which` / `type` as marco suggested to actually debug this and figure out the problem so it can be fixed upstream. But replacing `cd` with `command cd` would force using the built in.

Thank you for your help :+1: 
So i am going to use `command cd` instead of `cd` in all OSs(including macOs and ...) and use `type cd` before any usage of `command cd` in code for future debug in upstream.
Did i understand you right?

**Comment by MarcoFalke** [comment]
No, without knowing what the result (output) of `type cd` and `which cd` is, there is nothing we can do

**Comment by MarcoFalke** [comment]
https://travis-ci.org/bitcoin/bitcoin/jobs/601781884#L228

says

```
cd is a function

cd () 

{ 

    __zsh_like_cd cd "$@"

}

**Comment by ryanofsky** [comment]
> Anyone has any idea how to overwrite this `cd`? Can we use `pushd` maybe?

pushd won't help: 
https://github.com/rvm/rvm/blob/ad02fb2c7ae77c7fd92a4f68766a2c8140766d08/scripts/cd#L14-L16

I think we should debug this further (maybe turn on `set -x` or look at `type __zsh_like_cd` output) and figure out why the `__zsh_like_cd` is dying on mac with errexit, then report the issue to travis so it can be fixed, and in the meantime add a safe_cd function that either calls `command cd` or just toggles `errexit` before calling regular `cd`.

**Comment by ryanofsky** [comment]
Might be issue reported https://github.com/rvm/rvm/issues/4618#issuecomment-465145635 fixed by https://github.com/trinitronx/rvm/commit/bda58754a5ae59c99c08693f1215f5ae06c6b0f8

**Comment by MarcoFalke** [comment]
Good find @ryanofsky 

So upgrading to 1.29.9 (https://github.com/rvm/rvm/commit/8a3bf6a12f560c3b415224c5d6d6c195715fc59d) should fix this.

**Comment by Sjors** [comment]
`pushd` definitely doesn't help; I tried that before.

**Pull request closed without merging**